---
description: "Reports using `module` keyword instead of `namespace` for TypeScript namespaces."
title: "namespaceKeywords"
topic: "rules"
---

import { Tabs, TabItem } from "@astrojs/starlight/components";
import { RuleEquivalents } from "~/components/RuleEquivalents";
import RuleSummary from "~/components/RuleSummary.astro";

<RuleSummary plugin="ts" rule="namespaceKeywords" />

TypeScript originally used the `module` keyword to declare internal modules (namespaces).
The `namespace` keyword was later introduced to avoid confusion with ES6 modules.
Using `namespace` makes it clear you are defining a TypeScript namespace, not an ES6 module.

## Examples

<Tabs>
<TabItem label="❌ Incorrect">

```ts
module Values {
	export const value = 123;
}
```

```ts
declare module Utils {
	export function format(input: string): string;
}
```

</TabItem>
<TabItem label="✅ Correct">

```ts
namespace Values {
	export const value = 123;
}
```

```ts
declare namespace Utils {
	export function format(input: string): string;
}
```

```ts
// String literal module names are allowed (external/ambient module declarations)
declare module "external-module" {
	export const version: string;
}
```

</TabItem>
</Tabs>

## When Not To Use It

If you are working in a codebase that consistently uses `module`, and migrating to `namespace` is blocked on tooling issues or finding the time to do it, you may not want this rule.

You might consider using [Flint disable comments](/directives) and/or [configuration file disables](/configuration#files) for those specific situations instead of completely disabling this rule.

## Equivalents in Other Linters

<RuleEquivalents pluginId="ts" ruleId="namespaceKeywords" />
