---
description: "Reports non-null assertions on the left side of nullish coalescing operators."
title: "nonNullAssertedNullishCoalesces"
topic: "rules"
---

import { Tabs, TabItem } from "@astrojs/starlight/components";
import { RuleEquivalents } from "~/components/RuleEquivalents";
import RuleSummary from "~/components/RuleSummary.astro";

<RuleSummary plugin="ts" rule="nonNullAssertedNullishCoalesces" />

The nullish coalescing operator (`??`) is designed to handle `null` and `undefined` values by providing a default when either is encountered.
Using a non-null assertion (`!`) on the left side of `??` is contradictory: you're asserting the value is not `null` or `undefined` while simultaneously providing a fallback for when it is.

This pattern typically indicates programmer confusion about these operators, or copy-paste errors.
The non-null assertion is always redundant in this context and can be removed.

## Examples

<Tabs>
<TabItem label="❌ Incorrect">

```ts
declare const value: string | undefined;

value! ?? "default";
```

```ts
declare function getValue(): string | null;

getValue()! ?? "fallback";
```

```ts
declare const object: { property?: string };

object.property! ?? "missing";
```

</TabItem>
<TabItem label="✅ Correct">

```ts
declare const value: string | undefined;

value ?? "default";
```

```ts
declare function getValue(): string | null;

getValue() ?? "fallback";
```

```ts
declare const object: { property?: string };

object.property ?? "missing";
```

```ts
declare const object: { nested?: { value: string } };

object.nested!.value ?? "default";
```

</TabItem>
</Tabs>

## Options

This rule is not configurable.

## When Not To Use It

If your project is still onboarding to TypeScript's `strictNullChecks` compiler option, or more generally has unusual needs for null values, you might not receive false reports from this rule.
You might consider using [Flint disable comments](/directives) and/or [configuration file disables](/configuration#files) for those specific situations instead of completely disabling this rule.

## Further Reading

- [MDN: Nullish coalescing operator (??)](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing)
- [TypeScript: Non-null assertion operator](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#non-null-assertion-operator-postfix-)

## Equivalents in Other Linters

<RuleEquivalents pluginId="ts" ruleId="nonNullAssertedNullishCoalesces" />
