---
description: "Use a regular expression literal when the pattern is static."
title: "regexLiterals"
topic: "rules"
---

import { Tabs, TabItem } from "@astrojs/starlight/components";
import { RuleEquivalents } from "~/components/RuleEquivalents";
import RuleSummary from "~/components/RuleSummary.astro";

<RuleSummary plugin="ts" rule="regexLiterals" />

When the pattern and flags are static strings, using a regex literal is more concise and avoids unnecessary constructor calls.
Regex literals are also easier to read and cannot throw at runtime due to invalid patterns, since they are validated at parse time.

## Examples

<Tabs>
<TabItem label="❌ Incorrect">

```ts
const pattern = RegExp("abc");
```

```ts
const pattern = new RegExp("abc", "gi");
```

```ts
const pattern = RegExp("");
```

</TabItem>
<TabItem label="✅ Correct">

```ts
const pattern = /abc/;
```

```ts
const pattern = /abc/gi;
```

```ts
const pattern = /(?:)/;
```

```ts
// Dynamic patterns are fine
const pattern = new RegExp(userInput);
```

```ts
// Dynamic flags are fine
const pattern = new RegExp("abc", flags);
```

</TabItem>
</Tabs>

## Options

This rule is not configurable.

## When Not To Use It

If you prefer to always use `RegExp` constructors for consistency, or if you are building patterns from static pieces that are easier to read as strings, you might prefer to disable this rule.

## Further Reading

- [MDN: Regular expressions](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_expressions)
- [MDN: RegExp](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp)

## Equivalents in Other Linters

<RuleEquivalents pluginId="ts" ruleId="regexLiterals" />
