---
description: "Reports inconsistent letter casing in regex escape sequences."
title: "regexLetterCasing"
topic: "rules"
---

import { Tabs, TabItem } from "@astrojs/starlight/components";
import { RuleEquivalents } from "~/components/RuleEquivalents";
import RuleSummary from "~/components/RuleSummary.astro";

<RuleSummary plugin="ts" rule="regexLetterCasing" />

Consistent letter casing in regex escape sequences improves readability and makes patterns easier to compare.
This rule enforces:

- Lowercase for hexadecimal digits in `\x`, `\u`, and `\u{}` escape sequences
- Lowercase for literal letters in case-insensitive patterns (with the `i` flag)
- Uppercase for control character letters in `\c` escape sequences

## Examples

<Tabs>
<TabItem label="❌ Incorrect">

```ts
/\u000A/;
```

```ts
/\x0A/;
```

```ts
/\u{1F4A9}/u;
```

```ts
/\ca/u;
```

```ts
/Regexp/i;
```

```ts
/[A-Z]/i;
```

</TabItem>
<TabItem label="✅ Correct">

```ts
/\u000a/;
```

```ts
/\x0a/;
```

```ts
/\u{1f4a9}/u;
```

```ts
/\cA/u;
```

```ts
/regexp/i;
```

```ts
/[a-z]/i;
```

```ts
// Uppercase is fine without the i flag
/REGEXP/;
```

</TabItem>
</Tabs>

## Options

This rule is not configurable.

## When Not To Use It

If your codebase has an established convention of using uppercase hexadecimal digits in escape sequences, or if you prefer a different casing convention than what this rule enforces, you may want to disable it.

## Further Reading

- [MDN: Character escape](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Regular_expressions/Character_escape)
- [MDN: Unicode character class escape](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Regular_expressions/Unicode_character_class_escape)

## Equivalents in Other Linters

<RuleEquivalents pluginId="ts" ruleId="regexLetterCasing" />
