---
description: "Reports duplicate characters in regular expression character classes."
title: "regexDuplicateCharacterClassCharacters"
topic: "rules"
---

import { Tabs, TabItem } from "@astrojs/starlight/components";
import { RuleEquivalents } from "~/components/RuleEquivalents";
import RuleSummary from "~/components/RuleSummary.astro";

<RuleSummary plugin="ts" rule="regexDuplicateCharacterClassCharacters" />

Disallows duplicate characters in regular expression character classes.
When a character appears multiple times in a character class, only one occurrence is needed.
Duplicates are often typing mistakes and can be safely removed.

## Examples

### Duplicate Characters

<Tabs>
<TabItem label="❌ Incorrect">

```ts
const pattern = /[aaa]/;
```

</TabItem>
<TabItem label="✅ Correct">

```ts
const pattern = /[a]/;
```

</TabItem>
</Tabs>

### Character Already in Range

<Tabs>
<TabItem label="❌ Incorrect">

```ts
const pattern = /[a-z a]/;
```

</TabItem>
<TabItem label="✅ Correct">

```ts
const pattern = /[a-z]/;
```

</TabItem>
</Tabs>

### Duplicate Ranges

<Tabs>
<TabItem label="❌ Incorrect">

```ts
const pattern = /[a-z a-z]/;
```

</TabItem>
<TabItem label="✅ Correct">

```ts
const pattern = /[a-z]/;
```

</TabItem>
</Tabs>

## Options

This rule is not configurable.

## When Not To Use It

If you are intentionally using duplicate characters for documentation purposes or have auto-generated patterns, you might prefer to disable this rule.

## Further Reading

- [MDN: Character class](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Regular_expressions/Character_class)
- [MDN: Regular expressions](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_expressions)

## Equivalents in Other Linters

<RuleEquivalents
	pluginId="ts"
	ruleId="regexDuplicateCharacterClassCharacters"
/>
