---
description: "Reports indexed references in replacement strings that should use named references."
title: "regexNamedReplacements"
topic: "rules"
---

import { Tabs, TabItem } from "@astrojs/starlight/components";
import { RuleEquivalents } from "~/components/RuleEquivalents";
import RuleSummary from "~/components/RuleSummary.astro";

<RuleSummary plugin="ts" rule="regexNamedReplacements" />

When using `String.prototype.replace()` or `String.prototype.replaceAll()` with a regular expression that has named capturing groups, the replacement string should use named references (`$<name>`) instead of indexed references (`$1`, `$2`, etc.).

Named references make the code more self-documenting and less prone to errors when the regex pattern is modified.

## Examples

<Tabs>
<TabItem label="❌ Incorrect">

```ts
const result = "str".replace(/a(?<name>b)c/, "_$1_");
```

```ts
const result = "2024-01-15".replace(
	/(?<year>\d{4})-(?<month>\d{2})-(?<day>\d{2})/,
	"$2/$3/$1",
);
```

</TabItem>
<TabItem label="✅ Correct">

```ts
const result = "str".replace(/a(?<name>b)c/, "_$<name>_");
```

```ts
const result = "2024-01-15".replace(
	/(?<year>\d{4})-(?<month>\d{2})-(?<day>\d{2})/,
	"$<month>/$<day>/$<year>",
);
```

```ts
// Unnamed groups can use indexed references
const result = "str".replace(/a(b)c/, "_$1_");
```

</TabItem>
</Tabs>

## Options

This rule is not configurable.

## When Not To Use It

If you prefer the brevity of indexed replacements, or if your codebase has established conventions around using indexed references even for named groups, you might prefer to disable this rule.
Some developers find indexed references more familiar from other regex dialects.

## Further Reading

- [MDN: String.prototype.replace()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/replace)
- [MDN: Regular expressions - Named capturing group](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Regular_expressions/Named_capturing_group)

## Equivalents in Other Linters

<RuleEquivalents pluginId="ts" ruleId="regexNamedReplacements" />
