---
description: "Reports quantifiers `{0,}` in regular expressions that should use `*` instead."
title: "regexStarQuantifiers"
topic: "rules"
---

import { Tabs, TabItem } from "@astrojs/starlight/components";
import { RuleEquivalents } from "~/components/RuleEquivalents";
import RuleSummary from "~/components/RuleSummary.astro";

<RuleSummary plugin="ts" rule="regexStarQuantifiers" />

Reports quantifiers using `{0,}` in regular expressions and suggests using the more concise `*` quantifier instead.
Both forms are semantically equivalent (matching zero or more of the preceding element), but `*` is the idiomatic form.

## Examples

### Greedy Quantifier

<Tabs>
<TabItem label="❌ Incorrect">

```ts
const pattern = /a{0,}/;
```

</TabItem>
<TabItem label="✅ Correct">

```ts
const pattern = /a*/;
```

</TabItem>
</Tabs>

### Lazy Quantifier

<Tabs>
<TabItem label="❌ Incorrect">

```ts
const pattern = /a{0,}?/;
```

</TabItem>
<TabItem label="✅ Correct">

```ts
const pattern = /a*?/;
```

</TabItem>
</Tabs>

### Groups

<Tabs>
<TabItem label="❌ Incorrect">

```ts
const pattern = /(ab){0,}/;
```

</TabItem>
<TabItem label="✅ Correct">

```ts
const pattern = /(ab)*/;
```

</TabItem>
</Tabs>

### RegExp Constructor

<Tabs>
<TabItem label="❌ Incorrect">

```ts
const pattern = new RegExp("a{0,}");
```

</TabItem>
<TabItem label="✅ Correct">

```ts
const pattern = new RegExp("a*");
```

</TabItem>
</Tabs>

## Options

This rule is not configurable.

## When Not To Use It

If your project has a style guide that prefers the explicit `{0,}` syntax for clarity, or if you are programmatically generating regular expressions where `{0,}` is easier to produce, you might want to disable this rule.

## Further Reading

- [MDN: Quantifier](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_expressions/Quantifiers)
- [MDN: Regular expressions](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_expressions)

## Equivalents in Other Linters

<RuleEquivalents pluginId="ts" ruleId="regexStarQuantifiers" />
