---
description: "Reports negated character classes that can use shorthand escapes."
title: "regexConciseCharacterClassNegations"
topic: "rules"
---

import { Tabs, TabItem } from "@astrojs/starlight/components";
import { RuleEquivalents } from "~/components/RuleEquivalents";
import RuleSummary from "~/components/RuleSummary.astro";

<RuleSummary plugin="ts" rule="regexConciseCharacterClassNegations" />

Enforces using shorthand escape sequences like `\D`, `\W`, `\S`, and `\P` instead of negated character classes.
For example, `[^\d]` can be written more concisely as `\D`.

## Examples

### Negated Digit Character Class

<Tabs>
<TabItem label="❌ Incorrect">

```ts
const pattern = /[^\d]/;
```

</TabItem>
<TabItem label="✅ Correct">

```ts
const pattern = /\D/;
```

</TabItem>
</Tabs>

### Negated Word Character Class

<Tabs>
<TabItem label="❌ Incorrect">

```ts
const pattern = /[^\w]+/;
```

</TabItem>
<TabItem label="✅ Correct">

```ts
const pattern = /\W+/;
```

</TabItem>
</Tabs>

### Negated Whitespace Character Class

<Tabs>
<TabItem label="❌ Incorrect">

```ts
const pattern = /[^\s]/;
```

</TabItem>
<TabItem label="✅ Correct">

```ts
const pattern = /\S/;
```

</TabItem>
</Tabs>

### Negated Unicode Property Escape

<Tabs>
<TabItem label="❌ Incorrect">

```ts
const pattern = /[^\p{Letter}]/u;
```

</TabItem>
<TabItem label="✅ Correct">

```ts
const pattern = /\P{Letter}/u;
```

</TabItem>
</Tabs>

## Options

This rule is not configurable.

## When Not To Use It

If you prefer the explicit negation syntax `[^\d]` for readability, or if your codebase has an established convention using negated character classes, you might prefer to disable this rule.

## Further Reading

- [MDN: Character class escape](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Regular_expressions/Character_class_escape)
- [MDN: Unicode character class escape](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Regular_expressions/Unicode_character_class_escape)

## Equivalents in Other Linters

<RuleEquivalents pluginId="ts" ruleId="regexConciseCharacterClassNegations" />
