---
description: "Reports characters in regex character classes that appear as single visual characters but are made of multiple code points."
title: "regexMisleadingUnicodeCharacters"
topic: "rules"
---

import { Tabs, TabItem } from "@astrojs/starlight/components";
import { RuleEquivalents } from "~/components/RuleEquivalents";
import RuleSummary from "~/components/RuleSummary.astro";

<RuleSummary plugin="ts" rule="regexMisleadingUnicodeCharacters" />

Some characters that appear as a single visual unit are actually composed of multiple Unicode code points.
When these appear in regex character classes, each code point is matched separately, which is typically not the intended behavior.

This rule detects several types of multi-code-point characters in character classes:

- Surrogate pairs: Characters like ğŸ‘ that require two UTF-16 code units
- Combined characters: Base characters with combining marks like Ã (A + combining accent)
- Emoji with modifiers: Emoji with skin tone modifiers like ğŸ‘¶ğŸ»
- Regional indicator symbols: Flag emoji like ğŸ‡¯ğŸ‡µ (two regional indicators)
- ZWJ sequences: Characters joined with a zero-width joiner like ğŸ‘¨â€ğŸ‘©â€ğŸ‘¦

## Examples

<Tabs>
<TabItem label="âŒ Incorrect">

```ts
// Surrogate pair without unicode flag
const pattern = /[ğŸ‘]/;
```

```ts
// Combined character (A + combining accent)
const pattern = /[Ã]/;
```

```ts
// Emoji with skin tone modifier
const pattern = /[ğŸ‘¶ğŸ»]/u;
```

```ts
// Regional indicator symbols (flag)
const pattern = /[ğŸ‡¯ğŸ‡µ]/u;
```

```ts
// ZWJ sequence (family emoji)
const pattern = /[ğŸ‘¨â€ğŸ‘©â€ğŸ‘¦]/u;
```

</TabItem>
<TabItem label="âœ… Correct">

```ts
// Unicode flag handles surrogate pairs correctly
const pattern = /[ğŸ‘]/u;
```

```ts
// Match outside character class
const pattern = /ğŸ‘/;
```

```ts
// Use precomposed character
const pattern = /[Ã]/;
```

```ts
// Match emoji sequence outside character class
const pattern = /ğŸ‘¶ğŸ»/;
```

```ts
// Use \q{} syntax with v flag for grapheme clusters
const pattern = /[\q{ğŸ‘¶ğŸ»}]/v;
```

```ts
// Solo regional indicator is fine
const pattern = /[ğŸ‡¯]/u;
```

</TabItem>
</Tabs>

## Options

This rule is not configurable.

## When Not To Use It

If you intentionally want to match individual code points rather than visual characters, or if your regex pattern specifically needs to match partial Unicode sequences, you might prefer to disable this rule.
Some specialized text processing may require matching individual surrogate halves or combining marks.

## Further Reading

- [MDN: Regular expressions - Unicode character class escape](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Regular_expressions/Unicode_character_class_escape)
- [Unicode Standard Annex #29: Unicode Text Segmentation](https://unicode.org/reports/tr29/)
- [MDN: RegExp - Unicode flag](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/unicode)

## Equivalents in Other Linters

<RuleEquivalents pluginId="ts" ruleId="regexMisleadingUnicodeCharacters" />
