---
description: "Reports quantifiers that match exactly once, making them unnecessary."
title: "regexUnusedQuantifiers"
topic: "rules"
---

import { Tabs, TabItem } from "@astrojs/starlight/components";
import { RuleEquivalents } from "~/components/RuleEquivalents";
import RuleSummary from "~/components/RuleSummary.astro";

<RuleSummary plugin="ts" rule="regexUnusedQuantifiers" />

Reports quantifiers that match exactly once (`{1}` or `{1,1}`), which are unnecessary because the element already matches exactly once by default.

## Examples

### Exact Match of One

<Tabs>
<TabItem label="❌ Incorrect">

```ts
const pattern = /a{1}/;
```

</TabItem>
<TabItem label="✅ Correct">

```ts
const pattern = /a/;
```

</TabItem>
</Tabs>

### Range of One to One

<Tabs>
<TabItem label="❌ Incorrect">

```ts
const pattern = /a{1,1}/;
```

</TabItem>
<TabItem label="✅ Correct">

```ts
const pattern = /a/;
```

</TabItem>
</Tabs>

### Groups with Useless Quantifiers

<Tabs>
<TabItem label="❌ Incorrect">

```ts
const pattern = /(ab){1}/;
```

</TabItem>
<TabItem label="✅ Correct">

```ts
const pattern = /(ab)/;
```

</TabItem>
</Tabs>

### Character Classes

<Tabs>
<TabItem label="❌ Incorrect">

```ts
const pattern = /[a-z]{1}/;
```

</TabItem>
<TabItem label="✅ Correct">

```ts
const pattern = /[a-z]/;
```

</TabItem>
</Tabs>

## Options

This rule is not configurable.

## When Not To Use It

If you intentionally use `{1}` quantifiers for documentation or consistency purposes, you might want to disable this rule.

## Further Reading

- [MDN: Quantifier](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_expressions/Quantifiers)
- [MDN: Regular expressions](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_expressions)

## Equivalents in Other Linters

<RuleEquivalents pluginId="ts" ruleId="regexUnusedQuantifiers" />
