---
description: "Enforce consistent style for `.length` and `.size` checks."
title: "sizeComparisonOperators"
topic: "rules"
---

import { Tabs, TabItem } from "@astrojs/starlight/components";
import { RuleEquivalents } from "~/components/RuleEquivalents";
import RuleSummary from "~/components/RuleSummary.astro";

<RuleSummary plugin="ts" rule="sizeComparisonOperators" />

This rule enforces a consistent style for checking `.length` and `.size` properties in boolean contexts.

By default, the rule prefers implicit boolean coercion (e.g., `if (array.length)`) for its conciseness.
Alternatively, you can configure it to prefer explicit comparisons (e.g., `if (array.length > 0)`) for clarity.

## Examples

### Default (`style: "coercion"`)

<Tabs>
<TabItem label="❌ Incorrect">

```ts
declare const items: string[];
if (items.length > 0) {
}
```

```ts
declare const items: string[];
if (items.length === 0) {
}
```

```ts
declare const items: string[];
if (items.length !== 0) {
}
```

```ts
declare const mySet: Set<string>;
if (mySet.size > 0) {
}
```

</TabItem>
<TabItem label="✅ Correct">

```ts
declare const items: string[];
if (items.length) {
}
```

```ts
declare const items: string[];
if (!items.length) {
}
```

```ts
declare const items: string[];
items.length && doSomething();
```

```ts
declare const items: string[];
const hasItems = items.length ? "yes" : "no";
```

```ts
declare const mySet: Set<string>;
if (mySet.size) {
}
```

</TabItem>
</Tabs>

### With `style: "explicit"`

<Tabs>
<TabItem label="❌ Incorrect">

```ts
declare const items: string[];
if (items.length) {
}
```

```ts
declare const items: string[];
if (!items.length) {
}
```

```ts
declare const items: string[];
items.length && doSomething();
```

```ts
declare const items: string[];
const hasItems = items.length ? "yes" : "no";
```

```ts
declare const items: string[];
Boolean(items.length);
```

```ts
declare const mySet: Set<string>;
if (mySet.size) {
}
```

</TabItem>
<TabItem label="✅ Correct">

```ts
declare const items: string[];
if (items.length > 0) {
}
```

```ts
declare const items: string[];
if (items.length === 0) {
}
```

```ts
declare const items: string[];
if (items.length !== 0) {
}
```

```ts
declare const items: string[];
const count = items.length;
```

```ts
declare const items: string[];
const count = items.length ?? 0;
```

```ts
declare const items: string[];
const value = items.length || 1;
```

```ts
declare const mySet: Set<string>;
if (mySet.size > 0) {
}
```

</TabItem>
</Tabs>

## Options

### `style`

- **Type:** `"coercion" | "explicit"`
- **Default:** `"coercion"`

Which style to enforce:

- `"coercion"`: Prefer implicit boolean checks like `if (array.length)`
- `"explicit"`: Prefer explicit comparisons like `if (array.length > 0)`

## When Not To Use It

If your team doesn't have a preference for one style over the other, or if you want to allow both styles, you might prefer to disable this rule.

## Further Reading

- [MDN: Truthy](https://developer.mozilla.org/en-US/docs/Glossary/Truthy)
- [MDN: Falsy](https://developer.mozilla.org/en-US/docs/Glossary/Falsy)

## Equivalents in Other Linters

<RuleEquivalents pluginId="ts" ruleId="sizeComparisonOperators" />
